netmaster/master/api.go

```diff
@@ -296,3 +311,147 @@ func DeleteEndpointHandler(w http.ResponseWriter, r *http.Request, vars map[stri
 	// done. return resp
 	return delResp, nil
 }
+
+//ServiceProviderUpdateHandler handles service provider update event from netplugin
+func ServiceProviderUpdateHandler(w http.ResponseWriter, r *http.Request, vars map[string]string) (interface{}, error) {
+
+	var svcProvUpdReq SvcProvUpdateRequest
+
+	// Get object from the request
+	err := json.NewDecoder(r.Body).Decode(&svcProvUpdReq)
+
+	if err != nil {
+		log.Errorf("Error decoding ServiceUpdateRequest. Err %v", err)
+		return nil, err
+	}
+
+	log.Infof("Recieved ServiceProviderUpdate {%+v}", svcProvUpdReq)
+
+	stateDriver, err := utils.GetStateDriver()
+	if err != nil {
+		return nil, err
+	}
+
+	if svcProvUpdReq.Event == "start" {
+		//Received container start event from netplugin. Check if the Provider
+		//matches any service and perform service provider update if there is a matching
+		//service.
+
+		epCfg := &mastercfg.CfgEndpointState{}
+		epCfg.StateDriver = stateDriver
+		nwID := svcProvUpdReq.Network + "." + svcProvUpdReq.Tenant
+		epCfg.ID = getEpName(nwID, &intent.ConfigEP{Container: svcProvUpdReq.Container})
+
+		err = epCfg.Read(epCfg.ID)
+		if err != nil {
+			return nil, err
+		}
+
+		provider := &mastercfg.Provider{}
+		provider.IPAddress = svcProvUpdReq.IPAddress
+		provider.Tenant = svcProvUpdReq.Tenant
+		provider.Network = svcProvUpdReq.Network
+		provider.ContainerID = svcProvUpdReq.ContainerID
+		provider.Labels = make(map[string]string)
+
+		if epCfg.Labels == nil {
+			//endpoint cfg doesnt have labels
+			epCfg.Labels = make(map[string]string)
+		}
+
+		for k, v := range svcProvUpdReq.Labels {
+			provider.Labels[k] = v
+			epCfg.Labels[k] = v
+		}
+		//maintain the containerId in endpointstat for recovery
+		epCfg.ContainerID = svcProvUpdReq.ContainerID
+
+		err = epCfg.Write()
+		if err != nil {
+			log.Errorf("error writing ep config. Error: %s", err)
+			return nil, err
+		}
+
+		providerID := getProviderID(provider)
+		providerDbID := getProviderDbID(provider)
+		if providerID == "" || providerDbID == "" {
+			return nil, fmt.Errorf("Invalid ProviderID from providerInfo:{%v}", provider)
+		}
+
+		//update provider db
+		mastercfg.ProviderDb[providerDbID] = provider
```

Comment:
ah providerDbID is the container id. and it we wouldnt expect consecutive start events from the same container. And even if we do - i guess it would make sense to just the updated providerinfo.